### PR TITLE
Use the mode to aggregate category columns in overviews

### LINKED ISSUE #245: use mode when creating overviews for integer columns
when a table has an integer column the avg is being calculated when you usually need the mode (they are categorical values)

### PR DESCRIPTION
**Pull request by jgoizueta**
Fixes #245

### CODE REVIEW

**Comment by jgoizueta** [comment]
We have a related issue regarding missing stats: https://github.com/CartoDB/cartodb/issues/7190
We should check if stats are available and generate them if not, as in https://github.com/CartoDB/cartodb-postgresql/blob/master/scripts-available/CDB_Overviews.sql#L228

**Comment by jgoizueta** [comment]
@rafatower @iriberri please take a look

File: scripts-available/CDB_Overviews.sql [hunk 1]
```diff
@@ -531,6 +551,54 @@ AS $$
...
+RETURNS BOOLEAN
+AS $$
+DECLARE
+    schema_name TEXT;
+    table_name TEXT;
+    available BOOLEAN;
+    categorical BOOLEAN;
+BEGIN
+    SELECT * FROM _cdb_split_table_name(reloid) INTO schema_name, table_name;
+    SELECT n_distinct IS NOT NULL
```
Comment (ethervoid) [inline scripts-available/CDB_Overviews.sql]: So first check if the column is in the stats and if not you call analyze

File: scripts-available/CDB_Overviews.sql [hunk 2]
```diff
@@ -568,20 +637,30 @@ BEGIN
...
+      ELSE
+        RETURN Format('SUM(%s*%s)/%s::' || column_type, qualified_column, feature_count, total_feature_count);
+      END IF;
     END IF;
   WHEN 'text', 'character varying', 'character' THEN
-    IF _cdb_unlimited_text_column(reloid, column_name) THEN
-      -- TODO: this should not be applied to columns containing largish text;
-      -- it is intended only to short names/identifiers
-      RETURN  'CASE WHEN count(distinct ' || qualified_column || ') = 1 THEN MIN(' || qualified_column || ') WHEN ' || total_feature_count || ' < 5 THEN string_agg(distinct ' || qualified_column || ','' / '') ELSE ''*'' END::' || column_type;
+    IF _cdb_categorical_column(base_table, column_name) THEN
```
Comment (ethervoid) [inline scripts-available/CDB_Overviews.sql]: Why are you using the mode for text columns?

File: scripts-available/CDB_Overviews.sql [hunk 3]
```diff
@@ -568,20 +637,30 @@ BEGIN
...
+      RETURN Format('_cdb_mode(%s)::', qualified_column) || column_type;
     ELSE
-      RETURN 'CASE count(*) WHEN 1 THEN MIN(' || qualified_column || ') ELSE NULL END::' || column_type;
+      IF _cdb_unlimited_text_column(base_table, column_name) THEN
+        -- TODO: this should not be applied to columns containing largish text;
+        -- it is intended only to short names/identifiers
+        RETURN  'CASE WHEN count(distinct ' || qualified_column || ') = 1 THEN MIN(' || qualified_column || ') WHEN ' || total_feature_count || ' < 5 THEN string_agg(distinct ' || qualified_column || ','' / '') ELSE ''*'' END::' || column_type;
```
Comment (ethervoid) [inline scripts-available/CDB_Overviews.sql]: And here, what is the purpose?

File: scripts-available/CDB_Overviews.sql [hunk 4]
```diff
@@ -568,20 +637,30 @@ BEGIN
...
+      END IF;
     END IF;
   WHEN 'text', 'character varying', 'character' THEN
-    IF _cdb_unlimited_text_column(reloid, column_name) THEN
-      -- TODO: this should not be applied to columns containing largish text;
-      -- it is intended only to short names/identifiers
-      RETURN  'CASE WHEN count(distinct ' || qualified_column || ') = 1 THEN MIN(' || qualified_column || ') WHEN ' || total_feature_count || ' < 5 THEN string_agg(distinct ' || qualified_column || ','' / '') ELSE ''*'' END::' || column_type;
+    IF _cdb_categorical_column(base_table, column_name) THEN
+      RETURN Format('_cdb_mode(%s)::', qualified_column) || column_type;
     ELSE
```
Comment (ethervoid) [inline scripts-available/CDB_Overviews.sql]: Why don't you use elsif?

**Comment by ethervoid** [comment]
Looks good to me

File: scripts-available/CDB_Overviews.sql (hunk 4 above)
Comment (jgoizueta) [inline scripts-available/CDB_Overviews.sql]: Does it exist in PLPGSQL? I forgot...

File: scripts-available/CDB_Overviews.sql (hunk 1 above)
Comment (jgoizueta) [inline scripts-available/CDB_Overviews.sql]: If stats do no exist n_distinct will be NULL. Checking booleans is tricky with NULLs involved...

**Pull request merged**